collect_app/src/main/java/org/odk/collect/android/formentry/ODKView.java

```diff
@@ -480,39 +486,58 @@ public void scrollTo(@Nullable QuestionWidget qw) {
-                                    ExternalAppsUtils.asDecimalData(bundle.get(key)));
-                            break;
-                        default:
-                            throw new RuntimeException(
-                                    getContext().getString(R.string.ext_assign_value_error,
-                                            treeReference.toString(false)));
-                    }
 
-                    ((StringWidget) questionWidget).setDisplayValueFromModel();
-                    break;
+        if (bundle != null) {
+            Set<String> keys = bundle.keySet();
+            for (String key : keys) {
+                for (QuestionWidget questionWidget : widgets) {
+                    FormEntryPrompt prompt = questionWidget.getFormEntryPrompt();
+                    TreeReference treeReference =
+                            (TreeReference) prompt.getFormElement().getBind().getReference();
+
+                    if (treeReference.getNameLast().equals(key)) {
+                        switch (prompt.getDataType()) {
+                            case Constants.DATATYPE_TEXT:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asStringData(bundle.get(key)));
+                                ((StringWidget) questionWidget).setDisplayValueFromModel();
+                                break;
+                            case Constants.DATATYPE_INTEGER:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asIntegerData(bundle.get(key)));
+                                ((StringWidget) questionWidget).setDisplayValueFromModel();
+                                break;
+                            case Constants.DATATYPE_DECIMAL:
+                                formController.saveAnswer(prompt.getIndex(),
+                                        ExternalAppsUtils.asDecimalData(bundle.get(key)));
+                                ((StringWidget) questionWidget).setDisplayValueFromModel();
+                                break;
+                            case Constants.DATATYPE_BINARY:
+                                try {
+                                    Uri uri = (Uri) bundle.get(key);
+                                    try {
+                                        File destFile = FileUtils.createDestinationMediaFile(formController.getInstanceFile().getParent(), ContentResolverHelper.getFileExtensionFromUri(getContext(), uri));
+                                        //TODO might be better to use QuestionMediaManager in the future
+                                        FileUtils.saveAnswerFileFromUri(uri, destFile, getContext());
+                                        ((WidgetDataReceiver) questionWidget).setData(destFile);
+                                    } catch (SecurityException e) {
+                                        Timber.w(e);
```

Comment:
Will this implicitly clear any existing value?

I think there's one more error case missing -- if the type of the file passed from the external app doesn't match the question type. For example, a .mp3 file shouldn't be allowed for an image widget.

Reply:
When it comes to file formats please take a look below I described that it should be a separate pr if needed because we don't check file types using binary widgets normally so why here in intents group?

> Will this implicitly clear any existing value?

we call `setData()` in the last step so if any exception occurs it's just not called and doesn't remove an existing answer.

Reply:
> doesn't remove an existing answer

This is a weird case -- a user would have to get data from an external app, have it work fine, then get data again and have it fail. It's not very likely but then if it does happen I think it's nice for the answer to be cleared so it's easier for a user to see that something unexpected happened. It should be straightforward to clear in each `catch` block, right?

Reply:
Again like in case of checking file types I think it wouldn't align with what normal widgets outside intent groups does. If something goes wrong we don't remove an existed answer. We remove the old answer only in setData() so just before adding a new one where we are almost sure that everything is fine.

Reply:
Next time let's make sure we have these conversations when we're considering the spec and make sure the implementation is based on the spec! I'll file issues once this is merged.